tree/ntuple/v7/inc/ROOT/RNTupleProcessor.hxx

```diff
@@ -0,0 +1,293 @@
+*/
+// clang-format on
+class RNTupleProcessor {
+private:
+   // clang-format off
+   /**
+   \class ROOT::Experimental::RNTupleProcessor::RFieldContext
+   \ingroup NTuple
+   \brief Manager for a field as part of the RNTupleProcessor.
+
+   An RFieldContext contains two fields: a proto-field which is not connected to any page source but serves as the
+   blueprint for this particular field, and a concrete field that is connected to the page source currently connected
+   to the RNTupleProcessor object for reading. When a new page source is connected, the concrete field gets reset by
+   cloning the proto-field and connecting it to the new page source.
+
+   Apart from the fields themselves, the RFieldContext object also manages the pointer to the value for this particular
+   field that's read by the processor.
+   */
+   // clang-format on
+   class RFieldContext {
+   private:
+      std::unique_ptr<RFieldBase> fProtoField;
+      std::unique_ptr<RFieldBase> fConcreteField;
+      REntry::RFieldToken fToken;
+      std::shared_ptr<void> fValuePtr;
+
+   public:
+      RFieldContext(std::unique_ptr<RFieldBase> protoField, REntry::RFieldToken token)
+         : fProtoField(std::move(protoField)), fToken(token)
+      {
+      }
+
+      RFieldBase &GetProtoField() const { return *fProtoField; }
+      void ResetConcreteField() { fConcreteField = nullptr; }
+      void SetConcreteField() { fConcreteField = fProtoField->Clone(fProtoField->GetFieldName()); }
+      RFieldBase &GetConcreteField() const { return *fConcreteField; }
+      const REntry::RFieldToken &GetToken() const { return fToken; }
+      std::shared_ptr<void> GetValuePtr() const { return fValuePtr; }
+      void SetValuePtr(std::shared_ptr<void> ptr) { fValuePtr = ptr; }
+   };
+
+   // clang-format off
+   /**
+   \class ROOT::Experimental::RNTupleProcessor::RProcessorView
+   \ingroup NTuple
+   \brief View on the RNTupleProcessor iterator state.
+   */
+   // clang-format on
+   class RProcessorView {
+   private:
+      const REntry &fEntry;
+      const NTupleSize_t &fGlobalEntryIndex;
+      const NTupleSize_t &fLocalEntryIndex;
+      const std::size_t &fNTupleIndex;
+
+   public:
+      RProcessorView(const REntry &entry, const NTupleSize_t &globalEntryIndex, const NTupleSize_t &localEntryIndex,
+                     const std::size_t &ntupleIndex)
+         : fEntry(entry),
+           fGlobalEntryIndex(globalEntryIndex),
+           fLocalEntryIndex(localEntryIndex),
+           fNTupleIndex(ntupleIndex)
+      {
+      }
+
+      const REntry *operator->() const { return &fEntry; }
+      const REntry &GetEntry() const { return fEntry; }
+      NTupleSize_t GetGlobalEntryIndex() const { return fGlobalEntryIndex; }
+      NTupleSize_t GetLocalEntryIndex() const { return fLocalEntryIndex; }
+      std::size_t GetNTupleIndex() const { return fNTupleIndex; }
+   };
+
+   const std::vector<RNTupleSourceSpec> &fNTuples;
```

Comment:
A gap in my C++ knowledge :-). I think it makes sense to make it just a `const` here?

Reply:
Since this is only the specifications, i.e. light objects, I'd simply use a normal `std::vector<RNTupleSourceSpec> fNTuples` and accept the copy on construction.